deps/happy-eyeballs/happy-eyeballs.h

```diff
@@ -0,0 +1,126 @@
+/**
+ *	An implementation of RFC 6555 (Happy Eyeballs) to connect to IPv6 hosts with IPv4 fast fallback.
+ *	This implementation currently only works for TCP.
+ *
+ *	More precisely this algorithm will attempt the configured preferred protocol family, followed by
+ *	the other. So if the user has configured their operating system to prefer IPv4 it will attempt IPv4
+ *	first, followed by IPv6. Generally, though, the default is to prefer IPv6.
+ *
+ *	Initial implementation by james hurley <jamesghurley@gmail.com>
+ */
+
+#pragma once
+
+#ifdef _WIN32
+#include <winsock2.h>
+#include <ws2tcpip.h>
+#else
+#include <sys/socket.h>
+#include <unistd.h>
+#endif
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct happy_eyeballs_ctx;
+
+/**
+ *	Create and initialize a Happy Eyeballs session. On failure *context will always be NULL.
+ * 	You must call happy_eyeballs_destroy() to free any resources.
+ *	This function will return 0 on success and *context will be non-NULL.
+ *
+ *	-EINVAL if context is NULL
+ *	-ENOMEM if there is an allocation failure.
+ *  Another negative error code if there is a failure creating required resources.
+ */
+int happy_eyeballs_create(struct happy_eyeballs_ctx **context);
+
+/**
+ *	Connect to a provided hostname and port. This function will block for up to 1.2 seconds. After this function
+ *	returns with a success result (0), you need to check happy_eyeballs_try or happy_eyeballs_timedwait to
+ *	determine if a connection has been established.
+ *
+ *	Upon successfully starting the connection process and waiting for the initial delay timers, this function will
+ *	return 0 if the connection has been successfully established, or EAGAIN if the caller should wait longer.
+ */
+int happy_eyeballs_connect(struct happy_eyeballs_ctx *context,
+			   const char *hostname, int port);
+
+/**
+ *	Optionally set the interface address. You may pass 0 and NULL for these parameters, respectively, to clear
+ *	this setting. This must be called before happy_eyeballs_connect.
+ *
+ *	Returns 0 on success or -EINVAL if context is not set.
+ */
+int happy_eyeballs_set_bind_addr(struct happy_eyeballs_ctx *context,
+				 socklen_t addr_len,
+				 struct sockaddr_storage *addr_storage);
+
+/**
+ *	Fills socket_fd with a positive socket file descriptor if a winner is available.
+ *	Returns 0 if there is no socket available yet or -EINVAL if context is not set.
+ */
+int happy_eyeballs_get_socket_fd(struct happy_eyeballs_ctx *context);
+
+/**
+ *	Fills the addr struct with the winning socket address, if there is one.
```

Comment:
Maybe add comments to this and the other functions that should only be called after e.g. `happy_eyeballs_try` returns 0